### ISSUE TITLE
Separate stat masking(replication) from stat modifiers

### ISSUE THREAD

**Issue by moonshadow565** [LeagueSandbox/GameServer#492]
Currently stats implementation is incredibly messy and ureadable because stat modifiers try to use stat masks to add stat modfiers(?).
Because not all units have same stats and most of them don't even have them it feels realy hacky to client.

**Comment by MythicManiac** [comment]
- Replication here refers to copying a variable from the server to the client.
- We would need a system for marking variables for replication and automatically replicating on change.
- There are 7 different replicator types in total. Which are
  - AIProp
  - Barracks
  - BarracksDampener
  - Hero
  - Hq
  - Minion
  - Tower
- One way of implementing this is to use custom properties (e.g. `Replicate<int> someVar;`) which track the variable and auto-replicate on change.

----

if we could have this kind of behaviour it would be the best
```CS
class SomeObject
{
    Int someValue = 0;
};


var obj = new SomeObject;
List<object> tracking;
tracking.add(obj.someValue);
obj.someValue = 4;

print(tracking[0]); //this should print 4
```
— @moonshadow565 

----

```CS
class Replicate<T>
{
    //omitted...
    
    void OnValueChange() { DoSomething(); }
}

class SomeObject
{
    Replicate<int> someValue = 0;
}
```
should be sufficient

----

Proposed class structure
```CS
public enum ReplicatorType
{
    AIProp,
    Barracks,
    BarracksDampener,
    Hero,
    Hq,
    Minion,
    Tower
}
```
```CS
public abstract class BaseReplicator // Could be an interface as well
{
    protected abstract ReplicatorType _replicatorType
}

public virtual class BarracksReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.Barracks; }}
}

public virtual class BarracksDampenerReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.BarracksDampener; }}
}

public virtual class HeroReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.Hero; }}
}

public virtual class HqReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.Hq; }}
}

public virtual class MinionReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.Minion; }}
}

public virtual class TowerReplicator : BaseReplicator
{
    protected override ReplicatorType _replicatorType { get { return ReplicatorType.Tower; }}
}
```
```CS
public class Barracks(BarracksReplicator)
{
    [ReplicateField("mHP")]
    public Replicate<int> Health

    [ReplicateField("IsInvulnerable")]
    public Replicate<bool> IsInvulnerable

    [ReplicateField("mIsTargetable")]
    public Replicate<bool> IsTargetable

    [ReplicateField("mIsTargetableToTeamFlags")]
    public Replicate<byte> IsTargetableToTeamFlags
}
```
Replicate fields:
[420_docs.zip](https://github.com/LeagueSandbox/GameServer/files/1360794/420_docs.zip)

**Comment by moonshadow565** [comment]
FIxed in #570
